### PR TITLE
feat: support both mouse selection with the left mouse button and panning with the middle mouse button in the graph

### PR DESCRIPTION
**Pull request by aMoonkin**
### Description

support both mouse selection with the left mouse button and panning with the middle mouse button in the graph.

在画布中同时支持左键框选和中键 panning

#### Problem

Expected to support both mouse selection with the left mouse button and panning with the middle mouse button in the graph. But
1. x6-plugin-selection does not support selection with specified buttons.
2. x6 doesn't support only mouse wheel down panning

------

期待在在画布中同时支持左键框选和中键 panning
1.  x6-plugin-selection 不支持指定鼠标按键框选
2. panning 不支持仅鼠标中键移动

------

#### Changes

- added the "eventTypes" option to selection options, to specify the mouse buttons that triggers the interaction
- added "mouseWheelDown" to the "eventTypes" option of panning, to handle panning only when the mouse wheel is down.
- modified the selection example
------

- 为 selection 增加了 "eventTypes" 选项, 用于指定产生交互的鼠标按键
- 为 panning  选项中的 "eventTypes" 增加了 `mouseWheelDown`, 用于处理仅在鼠标滚轮点击时的移动行为
- 修改了 selection example

------

#### Concern

- In plugin-selection, when merging default configurations, using deep copy may not work as expected, e.g. passing `multipleSelectionModifiers: [],` in the selection configuration actually expects not to use any of the multiple selection buttons, but still makes the default configuration ['ctrl', 'meta'] take effect, so I used the following code to handle the behavior of `eventType` in my code. 'meta'] to take effect, so I handled the behavior of `eventType` in my code with the following code.

```javascript
if (!Array.isArray(this.options.eventTypes)) {
  // a new option, compatible with the behavior of older versions
  this.options.eventTypes = ['leftMouseDown', 'mouseWheelDown']
}
```

- The `leftMouseDown` of the `eventTypes` option in panning does not actually distinguish between the mouse down behavior of the left mouse button and the mouse wheel, whereas the `leftMouseDown` of the selection `eventTypes` option in panning only represents the mouse down behavior of the left mouse button. behavior of the left mouse button. Consider whether to accept this difference, and whether it is possible to support left mouse button and scroll wheel clicks independently on a larger scale.

------

- plugin-selection 中, merge 默认配置时, 使用深拷贝可能是不符合预期的, 比如在 selection 配置中传入`multipleSelectionModifiers: [],`, 实际是期待不使用任意一个多选按键, 但依然会使得默认配置 ['ctrl', 'meta'] 生效, 因此我在代码中用以下代码处理了 `eventType` 的行为.

```javascript
if (!Array.isArray(this.options.eventTypes)) {
  // a new option, compatible with the behavior of older versions
  this.options.eventTypes = ['leftMouseDown', 'mouseWheelDown']
}
```

- panning 中 `eventTypes` 选项的 `leftMouseDown` 实际没有区分鼠标左键和滚轮的 mouse down 行为, 而在本次新增的 selection `eventTypes` 选项中的 `leftMouseDown` 却仅代表鼠标左键的 mouse down 行为. 考虑是否接受这种差异性, 以及能否在更大规模的范围中独立支持鼠标左键和滚轮的点击行为.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Enhancement (changes that improvement of current feature or performance)
- [ ] Refactoring (changes that neither fixes a bug nor adds a feature)
- [ ] Test Case (changes that add missing tests or correct existing tests)
- [ ] Code style optimization (changes that do not affect the meaning of the code)
- [ ] Docs (changes that only update documentation)
- [ ] Chore (changes that don't modify src or test files)

### Self Check before Merge

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING**](https://github.com/antvis/x6/blob/master/CONTRIBUTING.md) document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: packages/x6-plugin-selection/src/index.ts [hunk 1]
```diff
@@ -47,6 +47,12 @@ export class Selection
       Selection.defaultOptions,
       options,
     )
+
+    if (!Array.isArray(this.options.eventTypes)) {
+      // a new option, compatible with the behavior of older versions
+      this.options.eventTypes = ['leftMouseDown', 'mouseWheelDown']
+    }
```
Comment (aMoonkin) [inline packages/x6-plugin-selection/src/index.ts:54]: due to unexpected `merge` method, I have to use this to keep the behavior works the same as before

File: packages/x6/src/graph/panning.ts [hunk 2]
```diff
@@ -195,7 +210,11 @@ export class PanningManager extends Base {
 }
 
 export namespace PanningManager {
-  type EventType = 'leftMouseDown' | 'rightMouseDown' | 'mouseWheel'
+  type EventType =
+    | 'leftMouseDown'
```
Comment (NewByVector) [inline packages/x6/src/graph/panning.ts:214]: 可以在文档上增加说明。

File: packages/x6-plugin-selection/src/index.ts [hunk 3]
```diff
@@ -304,21 +310,47 @@ export class Selection
...
+      this.graph.off('blank:mousedown', this.onBlankMouseWheelDown, this)
+    }
     this.graph.off('blank:click', this.onBlankClick, this)
     this.graph.off('cell:mousemove', this.onCellMouseMove, this)
     this.graph.off('cell:mouseup', this.onCellMouseUp, this)
     this.selectionImpl.off('box:mousedown', this.onBoxMouseDown, this)
   }
 
+  protected onBlankLeftMouseDown(mouseDownEvent: EventArgs['blank:mousedown']) {
+    if (mouseDownEvent.e.button !== 0) {
```
Comment (NewByVector) [inline packages/x6-plugin-selection/src/index.ts:339]:
```ts
if (mouseDownEvent.e.button === 0) {
  this.onBlankMouseDown(mouseDownEvent)
}
```
这样是不是更好

File: packages/x6-plugin-selection/src/index.ts [hunk 4]
```diff
@@ -449,9 +481,11 @@ export class Selection
 }
 
 export namespace Selection {
+  type SelectionEventType = 'leftMouseDown' | 'mouseWheelDown'
   export interface EventArgs extends SelectionImpl.EventArgs {}
   export interface Options extends SelectionImpl.CommonOptions {
     enabled?: boolean
+    eventTypes?: SelectionEventType[]
```
Comment (NewByVector) [inline packages/x6-plugin-selection/src/index.ts:488]: 同样需要在文档上说明

File: packages/x6-plugin-selection/src/index.ts [hunk 5]
```diff
@@ -47,6 +47,12 @@ export class Selection
       Selection.defaultOptions,
       options,
     )
+
+    if (!Array.isArray(this.options.eventTypes)) {
+      // a new option, compatible with the behavior of older versions
+      this.options.eventTypes = ['leftMouseDown', 'mouseWheelDown']
```
Comment (NewByVector) [inline packages/x6-plugin-selection/src/index.ts:53]: 在 `Selection.defaultOptions` 增加 `eventTypes` 的默认值更好。

**Review by NewByVector — CHANGES_REQUESTED** [review]

File: packages/x6-plugin-selection/src/index.ts (hunk 5 above)
Comment (aMoonkin) [reply packages/x6-plugin-selection/src/index.ts:53]: @NewByVector  加在 `Selection.defaultOptions` 上看起来更好. 但想要先确认 pr concern 部分的一点, 此处的 `merge` 是否应该改成浅拷贝?  以便把此处额外逻辑去掉.
```
this.options = ObjectExt.merge(
    { enabled: true },
    Selection.defaultOptions,
    options,
  )
```

Comment (NewByVector) [reply packages/x6-plugin-selection/src/index.ts:53]: 确实应该修改成浅拷贝👍。

File: packages/x6-plugin-selection/src/index.ts [hunk 6]
```diff
@@ -304,21 +304,45 @@ export class Selection
   }
 
   protected startListening() {
-    this.graph.on('blank:mousedown', this.onBlankMouseDown, this)
+    if (this.options.eventTypes?.includes('leftMouseDown')) {
```
Comment (NewByVector) [inline packages/x6-plugin-selection/src/index.ts:307]: 这里和下面同样，需要将判断逻辑放到事件回调中。

**Review by NewByVector — CHANGES_REQUESTED** [review]

File: packages/x6-plugin-selection/src/index.ts [hunk 7]
```diff
@@ -320,6 +320,14 @@ export class Selection
   }
 
   protected onBlankMouseDown({ e }: EventArgs['blank:mousedown']) {
+    const eventTypes = this.options.eventTypes
+    if (
+      !(eventTypes && eventTypes.includes('leftMouseDown') && e.button === 0) &&
```
Comment (NewByVector) [inline packages/x6-plugin-selection/src/index.ts:325]: 不好意思，这么晚才来 review 你的代码。十分抱歉。这里我感觉这样更合理：
```ts
protected allowBlankMouseDown() {
  const eventTypes = this.options.eventTypes
  return eventTypes.includes('leftMouseDown') && e.button === 0 || eventTypes.includes('mouseWheelDown') && e.button === 1
}
```

File: packages/x6/src/graph/panning.ts [hunk 8]
```diff
@@ -102,7 +102,10 @@ export class PanningManager extends Base {
 
   protected onMouseDown({ e }: { e: Dom.MouseDownEvent }) {
     const eventTypes = this.widgetOptions.eventTypes
-    if (!(eventTypes && eventTypes.includes('leftMouseDown'))) {
+    if (
+      !(eventTypes && eventTypes.includes('leftMouseDown') && e.button === 0) &&
```
Comment (NewByVector) [inline packages/x6/src/graph/panning.ts:106]: 同上

**Review by NewByVector — CHANGES_REQUESTED** [review]

File: packages/x6-plugin-selection/src/index.ts (hunk 7 above)
Comment (aMoonkin) [reply packages/x6-plugin-selection/src/index.ts:325]: 谢谢, 我周末修一下

**Comment by TalentJY** [comment]
请问 这个功能能支持动态修改交互的鼠标按键吗？ 比如我想要一开始默认是中键移动画布，然后想动态修改为左键和中键一起能移动等等 @aMoonkin

**Comment by aMoonkin** [comment]
> 请问 这个功能能支持动态修改交互的鼠标按键吗？ 比如我想要一开始默认是中键移动画布，然后想动态修改为左键和中键一起能移动等等 @aMoonkin

目前不支持, 理论来说好支持, Selection 暴露一个接口就可以了.

**Review by NewByVector — APPROVED** [review]

**Pull request merged**